src/glang/lib/std/unicode.c3

```diff
@@ -0,0 +1,412 @@
+    // 1. The maximal subpart at that offset is replaced by a single U+FFFD
+    // 2. The conversion proceeds at the offset immediately after the maximal subpart.
+
+    // Byte 1
+    if encoding_type == enum_utf8_invalid() {
+        return invalid_unicode_scalar();
+    }
+
+    if encoding_length == 1 {
+        return unsafe_decode_utf8_bytes({first_byte});
+    }
+
+    // Byte 2
+    if !self.underlying:has_next() {
+        return invalid_unicode_scalar(); // TODO: iterators should be nicer than this
+    }
+
+    let second_byte : u8 = self.underlying:peak();
+    if !is_2nd_utf8_byte_in_valid_range(encoding_type, second_byte) {
+        return invalid_unicode_scalar();
+    }
+
+    self.underlying:__builtin_get_next();
+    if encoding_length == 2 {
+        return unsafe_decode_utf8_bytes({first_byte, second_byte});
+    }
+
+    // Byte 3
+    if !self.underlying:has_next() {
+        return invalid_unicode_scalar();
+    }
+
+    let third_byte : u8 = self.underlying:peak();
+    if !is_trailing_utf8_byte_in_valid_range(third_byte) {
+        return invalid_unicode_scalar();
+    }
+
+    self.underlying:__builtin_get_next();
+    if encoding_length == 3 {
+        return unsafe_decode_utf8_bytes({first_byte, second_byte, third_byte});
+    }
+
+    // Byte 4
+    if !self.underlying:has_next() {
+        return invalid_unicode_scalar();
+    }
+
+    let forth_byte : u8 = self.underlying:peak();
+    if !is_trailing_utf8_byte_in_valid_range(forth_byte) {
+        return invalid_unicode_scalar();
+    }
+
+    self.underlying:__builtin_get_next();
+    return unsafe_decode_utf8_bytes({first_byte, second_byte, third_byte, forth_byte});
+}
+
+@implicit has_next : (self : UTF8DecodeIter mut&) -> bool = {
+    return self.underlying:has_next();
+}
+
+function encode_to_utf8 : (scalar : UnicodeScalar, buffer: Span<u8>) -> View<u8> = {
+    if is_in_range(scalar.value, 0x00, 0x7f) {
+        buffer:get(0) = Narrow<u8>(scalar.value & 0x7f);
+        return {&buffer.data, 1};
+    }
+
+    if is_in_range(scalar.value, 0x0080, 0x07ff) {
+        buffer:get(0) = Narrow<u8>(0xc0 | ((scalar.value >> 6) & 0x1f));
+        buffer:get(1) = Narrow<u8>(0x80 | (scalar.value & 0x3f));
+        return {&buffer.data, 2};
+    }
+
+    if is_in_range(scalar.value, 0x0800, 0x0fff)
+        or is_in_range(scalar.value, 0x1000, 0xcfff)
+        or is_in_range(scalar.value, 0xd000, 0xd7ff)
+        or is_in_range(scalar.value, 0xe000, 0xffff)
+    {
+        buffer:get(0) = Narrow<u8>(0xe0 | ((scalar.value >> 12) & 0x0f));
+        buffer:get(1) = Narrow<u8>(0x80 | ((scalar.value >> 6) & 0x3f));
+        buffer:get(2) = Narrow<u8>(0x80 | (scalar.value & 0x3f));
+        return {&buffer.data, 3};
+    }
+
+    if is_in_range(scalar.value, 0x10000, 0x3ffff)
+        or is_in_range(scalar.value, 0x40000, 0xfffff)
+        or is_in_range(scalar.value, 0x100000, 0x10ffff)
```

Comment:
Is `0x1f0000` invalid? Table 3-6 gives this pattern `000uuuuu zzzzyyyy yyxxxxxx`.

Reply:
UTF-16 has `2**16 + 2**20` codepoints (including the invalid surrogate pairs)... this needs 21 bits to represent in UTF-8 but anything past `0x10ffff` is invalid since it can't be represented in UTF-16.

Reply:
Ok, that makes sense